tests/IntegrityTests/ReferenceVersions.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System.Linq;
+using System.Xml;
+using System.Xml.Linq;
+using System.Xml.XPath;
+using NUnit.Framework;
+
+namespace IntegrityTests
+{
+    public class ReferenceVersions
+    {
+       [Test]
+        public void MustScopeReferencesToAMajorVersion()
+        {
+            new TestRunner("*.csproj", "Package References cannot have Version=\"*\" or restore will sometimes fail and yield old, incorrect, or mismatched versions.")
+                .Run(projectFilePath =>
+                {
+                    var xdoc = XDocument.Load(projectFilePath);
+                    var nsMgr = new XmlNamespaceManager(new NameTable());
+                    var query = "/Project/ItemGroup/PackageReference[@Version='*']";
+                    var xmlnsAtt = xdoc.Root.Attribute("xmlns");
+
+                    if (xmlnsAtt != null)
+                    {
+                        query = "/x:Project/x:ItemGroup/x:PackageReference[@Version='*']";
+                        var xmlns = xmlnsAtt.Value;
+                        nsMgr.AddNamespace("x", xmlns);
+                    }
+
+                    var badPackageRefs = xdoc.XPathSelectElements(query, nsMgr);
```

Comment:
Seems like it would be worth seeing if LINQ to XML would result in simpler code than all the stuff needed here to get xpath working.

Reply:
The LINQ-to-XML is roughly:

```
var xdoc = XDocument.Load(path);
XNamespace ns = xdoc.Root.Attribute("xmlns")?.Value ?? "";
 
var badRefs = xdoc
    .Descendants(ns + "PackageReference")
    .Where(pr => pr.Attribute("Version")?.Value == "*");
```

Save some lines, gain some confusing `?.` and ternary operators, and you're no longer validating that the PackageReference is inside `/Project/ItemGroup` ... it could be anywhere.

Unfortunately XML namespaces are awful and Linq-to-Xml can't get you out of them. :cry: